### PR TITLE
perf(opencypher): split on a literal delimiter without a regex

### PR DESCRIPTION
**Pull request by wkpark**
## What does this PR do?

`split(str, delimiter)` now finds the delimiter with `indexOf`. It no longer calls `str.split(Pattern.quote(delimiter), -1)`.

The result is unchanged:

- The string is cut at every non-overlapping occurrence of the delimiter, left to right.
- Empty leading, inner and trailing pieces are kept.
- The empty-delimiter branch (#5390) and the argument checks (#5798, #6608) are untouched.

## Motivation

The quoted pattern (`\Q...\E`) never qualifies for `String.split`'s fast path. Every call to `split()`, once per row, therefore compiled a `Pattern`, even for a one-character delimiter. On JDK 21, splitting `"a,b,c"` this way allocates about 1,040 bytes per call. The fast path allocates 232.

We run LightRAG on ArcadeDB, and it stores chunk ids joined with `<SEP>`. A 14,289-vertex scan that returns `split(e.source_id, '<SEP>')` went from 26 ms to 17 ms (median, Bolt, same data and server).

## Related issues

None. This is a small, self-contained hot-path change.

## Additional Notes

- **Equivalence test.** `CypherSplitLiteralDelimiterTest` compares the function's output with `str.split(Pattern.quote(d), -1)`, the previous definition. It covers:
  - regex-syntax delimiters (`.`, `|`, `.*`, `$`, `\Q`)
  - a self-overlapping delimiter (`aa` in `aaaaa`)
  - a delimiter longer than the input
  - delimiters at either end, and a string that is only the delimiter
  - an empty input
  - non-ASCII text and surrogate pairs
  - delimiters that start or end with a lone surrogate, which must not cut a pair of the string but still match an unpaired one (added after review)
- **Mutations.** The test fails when the empty trailing piece is dropped (5 of the first 14 cases), and when occurrences may overlap (7 of 14). The plain `indexOf` loop without the surrogate check fails 5 of the 21 cases.
- **Differential check.** On 3,000,000 random strings and delimiters built from letters and surrogate units, the result equals `str.split(Pattern.quote(d), -1)` in every case.
- **Tests run.** `mvn -pl engine test -Dtest='CypherSplitLiteralDelimiterTest,com/arcadedb/query/opencypher/*Split*Test,com/arcadedb/query/opencypher/*String*Test'`: 62 tests, 0 failures. The whole openCypher suite was also run with this change applied alongside other local changes: 5,611 tests, 1 failure, `Issue6216AlgoWorkKnobBoundsTest`, which runs out of memory under `-Xmx1500m` on an unmodified `main` too.
- **Unchanged return type.** The function still returns an unmodifiable `List<String>`, as `List.of(...)` did.

## Checklist

- [ ] I have run the build using `mvn clean package` command (not the full reactor: the `engine` module was built and tested)
- [x] My unit tests cover both failure and success scenarios

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Cypher string splitting with literal delimiters, including special characters, overlapping patterns, empty sections, Unicode, and emoji.
  * Corrected delimiter handling around Unicode surrogate pairs to avoid splitting within a character while still supporting unpaired surrogates.
  * Preserved leading, trailing, and consecutive empty pieces for consistent results.

* **Tests**
  * Added comprehensive coverage for literal delimiters, surrogate characters, Unicode text, and other edge cases.

### CODE REVIEW

File: engine/src/main/java/com/arcadedb/query/opencypher/executor/CypherSplitFunction.java [hunk 1]
```diff
@@ -72,6 +72,17 @@ public Object execute(final Object[] args, final CommandContext context) {
...
     }
 
-    return List.of(str.split(Pattern.quote(delimiter), -1));
+    // The delimiter is a literal, so it is found with indexOf. String.split(Pattern.quote(delimiter), -1) compiled a
+    // regex on every call, once per row: a quoted pattern never takes String.split's fast path, not even for a
+    // one-character delimiter. The pieces are the ones that call returns - every occurrence, left to right and not
+    // overlapping, with the empty leading, inner and trailing pieces kept.
+    final List<String> pieces = new ArrayList<>();
+    int from = 0;
+    for (int at = str.indexOf(delimiter); at >= 0; at = str.indexOf(delimiter, from)) {
```
[the remark this answers is not in the document]
Comment (wkpark) [reply engine/src/main/java/com/arcadedb/query/opencypher/executor/CypherSplitFunction.java]: Confirmed, and fixed in a87a2a308.

- **Reproduced:** on JDK 21, a lone `\uD83D` or `\uDE00` delimiter split `"😀"` into two pieces with the `indexOf` loop. The quoted regex returns one piece.
- **Fix:** kept the literal scan. An occurrence is now skipped when the delimiter starts with a low surrogate preceded by a high surrogate in the string, or ends with a high surrogate followed by a low one. Those are the only ways a UTF-16 match can land inside a pair. An unpaired surrogate in the string still matches, as it does for the regex.
- **Tests:** seven lone-surrogate cases were added to `CypherSplitLiteralDelimiterTest`. Five of them fail on the previous loop. The test class and its neighbours now run 62 tests with 0 failures.
- **Differential check:** 3,000,000 random strings and delimiters over `a`, `b` and two surrogate pairs' units, compared with `str.split(Pattern.quote(d), -1)`. The new code agrees on all of them. The previous loop differed on 122,809.

**Comment by lvca** [comment]
@claude review this

**Comment by wkpark** [comment]
The red lanes on a87a2a308 are not caused by this change. It touches only `CypherSplitFunction` and its test. a87a2a308 has the same tree as 913b8b253: the only difference is that the commits were re-signed.

| Lane | Failure | Why it is unrelated |
|---|---|---|
| `unit-tests` | `Issue5596MergeCoverageTest.aCleanAppendConflictIsStillMerged:314` (1 of 15,233) | It passed on the same tree in run 35172124075 (913b8b253) and on 2a1fcde7d. |
| `slow-unit-tests` | `Issue6534QuiesceResizeRaceTest` (60 s timeout), `Issue6657CloseTimeRebuildPendingStatTest` | It passed on 2a1fcde7d. `unit-tests` and `slow-unit-tests` are also red on current `main` pushes (e.g. 35168051543). |
| `js-e2e-tests` | `URL /api/v1/ready not accessible after 120000ms`: the server container never became ready | It passed on the same tree in 35172124075. |
| `ha-integration-tests` | | It is red on every recent run, `main` included. |
| `coverage-report` | `CodacyCoverageReporter: Either a project or account API token must be provided` | The token secret is not exposed to fork PRs. |
| `claude-review` | `Could not fetch an OIDC token` | OIDC is not available to fork PRs. |

`build-and-package`, `lint`, `opencypher-tck-tests`, `integration-tests`, `vector-unit-tests` and the other e2e lanes are green.

On the non-blocking nit in the Claude review: I left the boundary check inline to keep the diff small. I can extract it into `wouldSplitSurrogatePair(str, at, end)` if you prefer.

**Comment by lvca** [comment]
Thanks @wkpark !

**Comment by lvca** [comment]
Merged. Thanks for the perf work and for chasing down the surrogate-pair edge case CodeRabbit flagged - the fix and the added test cases are correct and well covered.

One small readability tweak (the inline surrogate-pair check → a named `splitsSurrogatePair()` helper) landed separately in 627e0d5ac, since edits couldn't be pushed to this branch (org-owned forks don't support maintainer edits). No behavior change, same tests still pass.

---
❤️ Thanks for helping make ArcadeDB better! ArcadeDB is free and open source, sustained by its community. If it's useful to you or your company, please consider [becoming a sponsor](https://github.com/sponsors/ArcadeData) to keep development going.
